File: java/17-jammy-ubuntu-22.04-fips/diagnostics/test-images/basic-test-image/src/main/TlsTestSuite.java
/* TlsTestSuite.java
 *
 * Copyright (C) 2006-2025 wolfSSL Inc.
 *
 * This file is part of wolfSSL.
 *
 * wolfSSL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * wolfSSL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
 */
import java.io.*;
import java.net.*;
import java.util.*;
import java.security.*;
import java.security.cert.*;
import javax.net.ssl.*;
import javax.net.ssl.SNIHostName;

/**
 * SSL/TLS test suite for wolfSSL Java FIPS image.
 * This class demonstrates usage of JSSE APIs with the wolfJSSE provider
 * and verifies TLS connectivity to various public endpoints.
 */
public class TlsTestSuite {

    private static final String PROVIDER_WOLFJSSE = "wolfJSSE";

    /* Test endpoints - public HTTPS sites for connectivity testing */
    private static final String[] TEST_ENDPOINTS = {
        "www.google.com:443",
        "www.wolfssl.com:443",
        "httpbin.org:443"
    };

    public static void main(String[] args) {
        TlsTestSuite suite = new TlsTestSuite();
        suite.runAllTests();
    }

    public void runAllTests() {
        System.out.println("=== wolfSSL FIPS SSL/TLS Test Suite ===\n");

        /* Show system properties and environment variables */
        System.out.println("Debug configuration:");
        System.out.println("  wolfjce.debug system property: " +
            System.getProperty("wolfjce.debug"));
        System.out.println("  wolfjsse.debug system property: " +
            System.getProperty("wolfjsse.debug"));
        System.out.println("  WOLFJCE_DEBUG env var: " +
            System.getenv("WOLFJCE_DEBUG"));
        System.out.println("  WOLFJSSE_DEBUG env var: " +
            System.getenv("WOLFJSSE_DEBUG"));

        /* Manually set debug properties if environment variables are set */
        if ("true".equals(System.getenv("WOLFJCE_DEBUG"))) {
            System.setProperty("wolfjce.debug", "true");
            System.out.println("  Manually enabled wolfjce.debug from env var");
        }
        if ("true".equals(System.getenv("WOLFJSSE_DEBUG"))) {
            System.setProperty("wolfjsse.debug", "true");
            System.out.println("  Manually enabled wolfjsse.debug from env var");
        }
        System.out.println();

        try {
            /* Verify provider setup */
            verifyProviderSetup();

            /* Test SSLContext creation */
            testSslContextCreation();

            /* Test TLS connections */
            testTlsConnections();

            /* Test SSL socket creation */
            testSslSocketCreation();

            /* Test certificate validation */
            testCertificateValidation();

            /* Test TLS protocol versions */
            testTlsProtocolVersions();

            /* Test cipher suite configuration */
            testCipherSuites();

            System.out.println("\n=== All SSL/TLS Tests PASSED ===");

        } catch (Exception e) {
            System.err.println("\nERROR: SSL/TLS test failed!");
            System.err.println("Exception: " + e.getClass().getName() +
                ": " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private void verifyProviderSetup()
        throws SecurityException {

        System.out.println("Verifying wolfJSSE Provider Setup:");

        Provider wolfJSSE = Security.getProvider(PROVIDER_WOLFJSSE);
        if (wolfJSSE == null) {
            throw new SecurityException("wolfJSSE provider not found");
        }

        System.out.println("   wolfJSSE provider found: " +
            wolfJSSE.getName() + " v" + wolfJSSE.getVersion());
        System.out.println("   Provider info: " + wolfJSSE.getInfo());
        System.out.println();
    }

    private void testSslContextCreation()
        throws Exception {

        System.out.println("Testing SSLContext Creation:");

        String[] protocols = {"TLS", "TLSv1.2", "TLSv1.3", "DEFAULT"};

        for (String protocol : protocols) {
            try {
                SSLContext context = SSLContext.getInstance(protocol);

                /* Verify using wolfJSSE */
                if (!PROVIDER_WOLFJSSE.equals(
                        context.getProvider().getName())) {
                    throw new SecurityException("SSLContext " + protocol +
                        " not using wolfJSSE provider");
                }

                /* Initialize with default trust managers */
                context.init(null, null, null);

                System.out.println("   SSLContext " + protocol +
                    ": Created and initialized (wolfJSSE)");

            } catch (NoSuchAlgorithmException e) {
                System.out.println("   - SSLContext " + protocol +
                    ": Not available");
            }
        }
        System.out.println();
    }

    private void testTlsConnections()
        throws Exception {

        System.out.println("Testing TLS Connections to Public Endpoints:");
        System.out.println(
            "   Using WKS system cacerts for certificate validation");

        /* Load the WKS system cacerts explicitly */
        String javaHome = System.getProperty("java.home");
        String cacertsPath = javaHome + "/lib/security/cacerts";

        KeyStore trustStore = KeyStore.getInstance("WKS");
        try (FileInputStream fis = new FileInputStream(cacertsPath)) {
            trustStore.load(fis, "changeitchangeit".toCharArray());
            System.out.println("   Loaded " + trustStore.size() +
                " CA certificates from WKS cacerts");
        }

        /* Create TrustManagerFactory with our WKS cacerts */
        TrustManagerFactory tmf = TrustManagerFactory.getInstance("PKIX");
        tmf.init(trustStore);

        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, tmf.getTrustManagers(), null);
        SSLSocketFactory factory = context.getSocketFactory();

        int failureCount = 0;
        Exception lastException = null;

        for (String endpoint : TEST_ENDPOINTS) {
            String[] parts = endpoint.split(":");
            String host = parts[0];
            int port = Integer.parseInt(parts[1]);

            try {
                System.out.println("   Testing connection to " +
                    endpoint + "...");

                SSLSocket socket = (SSLSocket) factory.createSocket(host, port);
                socket.setSoTimeout(10000); /* 10 second timeout */

                /* Set SNI (Server Name Indication) for proper virtual
                 * host handling */
                SSLParameters sslParams = socket.getSSLParameters();
                sslParams.setServerNames(Arrays.asList(new SNIHostName(host)));
                socket.setSSLParameters(sslParams);

                /* Perform TLS handshake */
                socket.startHandshake();

                /* Get session information */
                SSLSession session = socket.getSession();
                System.out.println("     TLS handshake successful");
                System.out.println("     Protocol: " + session.getProtocol());
                System.out.println("     Cipher Suite: " +
                    session.getCipherSuite());
                System.out.println("     Peer certificates: " +
                    session.getPeerCertificates().length);

                /* Show certificate chain details */
                inspectCertificateChain(
                    session.getPeerCertificates(), "     ");

                /* Test basic HTTP request */
                testHttpsRequest(socket, host);

                socket.close();
                System.out.println("     Connection closed successfully");

            } catch (Exception e) {
                failureCount++;
                lastException = e;
                String message = e.getMessage();
                System.out.println("     Connection failed: " + message);

                /* Categorize the type of failure */
                if (message != null) {
                    if (message.contains("error code: -140")) {
                        System.out.println("     ASN.1 parsing error: " +
                            "Certificate uses unsupported " +
                            "algorithms/structures");
                        System.out.println("     This may indicate " +
                            "Ed25519/Ed448 signatures or other non-FIPS " +
                            "algorithms");
                        System.out.println("     wolfSSL FIPS build may not " +
                            "include required ASN.1 parsers");
                    } else if (message.contains("error code: -275")) {
                        System.out.println("     Self-signed certificate " +
                            "error: May indicate SNI or chain validation " +
                            "issues");
                        System.out.println("     Check if proper certificate " +
                            "is being served for this hostname");
                    } else if (message.contains("certificate") ||
                               message.contains("ASN") ||
                        message.contains("error code: -188")) {
                        System.out.println("     Certificate validation " +
                            "failed");
                        System.out.println("     This may indicate " +
                            "certificate chain or validation issues");
                        System.out.println("     wolfJSSE TLS stack is " +
                            "working but certificate validation failed");
                    } else if (message.contains("Connection refused") ||
                               message.contains("timeout")) {
                        System.out.println("     Network connectivity issue");
                    } else {
                        System.out.println("     Other SSL/TLS issue: " +
                            e.getClass().getSimpleName());
                    }
                } else {
                    System.out.println("     SSL/TLS connection failed");
                }
                System.out.println("     wolfJSSE provider is functioning " +
                    "for TLS operations");

                /* Try to get certificate chain details even when
                 * validation fails */
                System.out.println("     Attempting to retrieve certificate " +
                    "chain for analysis...");
                inspectCertificateChainOnFailure(host, port);
            }
        }

        /* Check if any connections failed */
        if (failureCount > 0) {
            System.out.println("   FAILURE SUMMARY:");
            System.out.println("   " + failureCount + " out of " +
                TEST_ENDPOINTS.length + " TLS connections failed");
            throw new SecurityException("TLS connection failures. " +
                "Last error: " + (lastException != null ?
                    lastException.getMessage() : "Unknown"));
        }

        System.out.println("   All TLS connections successful - FIPS " +
            "certificate validation working correctly");
        System.out.println();
    }

    private void testHttpsRequest(SSLSocket socket, String host)
        throws Exception {

        /* Send a simple HTTP GET request */
        PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
        BufferedReader in = new BufferedReader(
            new InputStreamReader(socket.getInputStream()));

        out.println("GET / HTTP/1.1");
        out.println("Host: " + host);
        out.println("Connection: close");
        out.println();

        /* Read response headers */
        String line;
        boolean foundStatusLine = false;
        while ((line = in.readLine()) != null && !line.isEmpty()) {
            if (line.startsWith("HTTP/")) {
                System.out.println("     HTTP Response: " + line);
                foundStatusLine = true;
                break;
            }
        }

        if (!foundStatusLine) {
            throw new Exception("No valid HTTP response received");
        }
    }

    private void testSslSocketCreation()
        throws Exception {

        System.out.println("Testing SSL Socket Creation:");

        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, null, null);
        SSLSocketFactory factory = context.getSocketFactory();

        /* Test socket creation without connection */
        SSLSocket socket = (SSLSocket) factory.createSocket();

        System.out.println("   SSL Socket created successfully");
        System.out.println("   Supported protocols: " +
            Arrays.toString(socket.getSupportedProtocols()));
        System.out.println("   Enabled protocols: " +
            Arrays.toString(socket.getEnabledProtocols()));
        System.out.println("   Supported cipher suites: " +
            socket.getSupportedCipherSuites().length + " available");

        socket.close();
        System.out.println();
    }

    private void testCertificateValidation()
        throws Exception {

        System.out.println("Testing Certificate Validation:");

        /* Load WKS system cacerts explicitly */
        String javaHome = System.getProperty("java.home");
        String cacertsPath = javaHome + "/lib/security/cacerts";

        KeyStore trustStore = KeyStore.getInstance("WKS");
        try (FileInputStream fis = new FileInputStream(cacertsPath)) {
            trustStore.load(fis, "changeitchangeit".toCharArray());
            System.out.println("   Loaded " + trustStore.size() +
                " CA certificates from WKS cacerts");
        }

        /* Test trust manager factory with WKS cacerts */
        TrustManagerFactory tmf = TrustManagerFactory.getInstance("PKIX");

        /* Verify it's using wolfJSSE */
        if (!PROVIDER_WOLFJSSE.equals(tmf.getProvider().getName())) {
            throw new SecurityException(
                "TrustManagerFactory not using wolfJSSE provider");
        }

        tmf.init(trustStore);
        TrustManager[] trustManagers = tmf.getTrustManagers();

        System.out.println("   TrustManagerFactory created with wolfJSSE");
        System.out.println("   Trust managers initialized with WKS cacerts: " +
            trustManagers.length + " managers");

        /* Test key manager factory */
        KeyManagerFactory kmf = KeyManagerFactory.getInstance("PKIX");

        /* Verify it's using wolfJSSE */
        if (!PROVIDER_WOLFJSSE.equals(kmf.getProvider().getName())) {
            throw new SecurityException(
                "KeyManagerFactory not using wolfJSSE provider");
        }

        System.out.println("   KeyManagerFactory created with wolfJSSE");
        System.out.println();
    }

    private void testTlsProtocolVersions()
        throws Exception {

        System.out.println("Testing TLS Protocol Versions:");

        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, null, null);
        SSLSocketFactory factory = context.getSocketFactory();
        SSLSocket socket = (SSLSocket) factory.createSocket();

        String[] supportedProtocols = socket.getSupportedProtocols();
        String[] enabledProtocols = socket.getEnabledProtocols();

        System.out.println("   Supported TLS protocols:");
        for (String protocol : supportedProtocols) {
            System.out.println("     - " + protocol);
        }

        System.out.println("   Enabled TLS protocols:");
        for (String protocol : enabledProtocols) {
            System.out.println("     " + protocol);
        }

        /* Verify TLS 1.2 and 1.3 are supported */
        boolean tls12Supported =
            Arrays.asList(supportedProtocols).contains("TLSv1.2");
        boolean tls13Supported =
            Arrays.asList(supportedProtocols).contains("TLSv1.3");

        if (!tls12Supported) {
            throw new SecurityException("TLS 1.2 not supported");
        }
        System.out.println("   TLS 1.2 support verified");

        if (tls13Supported) {
            System.out.println("   TLS 1.3 support verified");
        }

        socket.close();
        System.out.println();
    }

    private void testCipherSuites()
        throws Exception {

        System.out.println("Testing Cipher Suites:");

        SSLContext context = SSLContext.getInstance("TLS");
        context.init(null, null, null);
        SSLSocketFactory factory = context.getSocketFactory();
        SSLSocket socket = (SSLSocket) factory.createSocket();

        String[] supportedCipherSuites =
            socket.getSupportedCipherSuites();
        String[] enabledCipherSuites =
            socket.getEnabledCipherSuites();

        System.out.println("   Total supported cipher suites: " +
            supportedCipherSuites.length);
        System.out.println("   Total enabled cipher suites: " +
            enabledCipherSuites.length);

        /* Count FIPS-approved cipher suites */
        int fipsAesCipherCount = 0;
        int fipsEcdheCipherCount = 0;

        System.out.println("   Enabled FIPS-approved cipher suites:");
        for (String cipherSuite : enabledCipherSuites) {
            if (cipherSuite.contains("AES")) {
                fipsAesCipherCount++;
                System.out.println("     " + cipherSuite);
            } else if (cipherSuite.contains("ECDHE")) {
                fipsEcdheCipherCount++;
                System.out.println("     " + cipherSuite);
            }
        }

        if (fipsAesCipherCount == 0) {
            throw new SecurityException("No AES cipher suites enabled");
        }

        System.out.println("   FIPS-approved AES cipher suites available: " +
            fipsAesCipherCount);
        System.out.println("   ECDHE cipher suites available: " +
            fipsEcdheCipherCount);

        socket.close();
        System.out.println();
    }

    /**
     * Inspect certificate chain details for debugging purposes
     */
    private void inspectCertificateChain(
        java.security.cert.Certificate[] certChain, String indent) {

        System.out.println(indent + "Certificate chain details:");

        for (int i = 0; i < certChain.length; i++) {
            if (certChain[i] instanceof X509Certificate) {
                X509Certificate cert = (X509Certificate) certChain[i];
                System.out.println(indent + "  [" + i + "] Subject: " +
                    cert.getSubjectX500Principal().getName());
                System.out.println(indent + "      Issuer: " +
                    cert.getIssuerX500Principal().getName());
                System.out.println(indent + "      Signature Algorithm: " +
                    cert.getSigAlgName());
                System.out.println(indent + "      Public Key Algorithm: " +
                    cert.getPublicKey().getAlgorithm());
                System.out.println(indent + "      Valid From: " +
                    cert.getNotBefore());
                System.out.println(indent + "      Valid To: " +
                    cert.getNotAfter());

                /* Check for extensions that might cause issues */
                try {
                    boolean[] keyUsage = cert.getKeyUsage();
                    if (keyUsage != null) {
                        System.out.println(indent + "      Key Usage: " +
                            Arrays.toString(keyUsage));
                    }

                    /* Check basic constraints */
                    int basicConstraints = cert.getBasicConstraints();
                    if (basicConstraints >= 0) {
                        System.out.println(indent +
                            "      Basic Constraints: CA=true, Path Length=" +
                            basicConstraints);
                    } else if (basicConstraints == -1) {
                        System.out.println(indent +
                            "      Basic Constraints: CA=false (end entity)");
                    }
                } catch (Exception e) {
                    System.out.println(indent +
                        "      Extension analysis failed: " + e.getMessage());
                }
            } else {
                System.out.println(indent + "  [" + i +
                    "] Non-X509 certificate: " +
                    certChain[i].getClass().getName());
            }
        }
    }

    /**
     * Attempt to retrieve certificate chain when validation fails,
     * using a permissive trust manager
     */
    private void inspectCertificateChainOnFailure(String host, int port) {
        try {
            /* Create a permissive trust manager that accepts
             * all certificates */
            TrustManager[] permissiveTrustManagers = new TrustManager[] {
                new X509TrustManager() {
                    public X509Certificate[] getAcceptedIssuers() {
                        return new X509Certificate[0];
                    }
                    public void checkClientTrusted(X509Certificate[] certs,
                        String authType) {
                        /* Accept all client certificates for inspection */
                    }
                    public void checkServerTrusted(X509Certificate[] certs,
                        String authType) {
                        /* Accept all server certificates for inspection */
                        System.out.println("       Retrieved " + certs.length +
                            " certificates for analysis");
                        inspectCertificateChain(certs, "       ");
                    }
                }
            };

            /* Create SSL context with permissive trust manager */
            SSLContext permissiveContext = SSLContext.getInstance("TLS");
            permissiveContext.init(null, permissiveTrustManagers, null);

            SSLSocketFactory permissiveFactory =
                permissiveContext.getSocketFactory();
            SSLSocket permissiveSocket =
                (SSLSocket) permissiveFactory.createSocket(host, port);
            permissiveSocket.setSoTimeout(5000);

            /* Set SNI for proper certificate retrieval */
            SSLParameters sslParams = permissiveSocket.getSSLParameters();
            sslParams.setServerNames(Arrays.asList(new SNIHostName(host)));
            permissiveSocket.setSSLParameters(sslParams);

            /* Perform handshake (should succeed with permissive
             * trust manager) */
            permissiveSocket.startHandshake();

            permissiveSocket.close();

        } catch (Exception inspectionException) {
            System.out.println("       Certificate chain inspection failed: " +
                inspectionException.getMessage());
        }
    }
}

File: node/18.20.8-bookworm-slim-fips/diagnostics/test-images/basic-test-image/src/tls_test_suite.js
#!/usr/bin/env node
/**
 * TLS/SSL Test Suite
 * Tests FIPS-approved TLS connections in a user application context
 */

const https = require('https');
const tls = require('tls');

class TlsTestSuite {
    constructor() {
        this.totalTests = 6;
        this.passedTests = 0;
        this.failedTests = 0;
    }

    logTest(name, passed, details = '') {
        const symbol = passed ? '✓' : '✗';
        console.log(`  ${symbol} ${name}: ${passed ? 'PASS' : 'FAIL'}`);
        if (details) {
            console.log(`    ${details}`);
        }

        if (passed) {
            this.passedTests++;
        } else {
            this.failedTests++;
        }
    }

    async test_https_connection() {
        return new Promise((resolve) => {
            try {
                const options = {
                    hostname: 'www.google.com',
                    port: 443,
                    path: '/',
                    method: 'HEAD',
                    timeout: 10000
                };

                const req = https.request(options, (res) => {
                    const cipher = res.socket.getCipher();
                    const protocol = res.socket.getProtocol();

                    const success = res.statusCode >= 200 && res.statusCode < 400;
                    this.logTest('HTTPS connection', success,
                        `${protocol} with ${cipher.name}`);

                    res.on('data', () => {});
                    res.on('end', () => resolve(success));
                });

                req.on('error', (error) => {
                    this.logTest('HTTPS connection', false, error.message);
                    resolve(false);
                });

                req.on('timeout', () => {
                    req.destroy();
                    this.logTest('HTTPS connection', false, 'Connection timeout');
                    resolve(false);
                });

                req.end();
            } catch (error) {
                this.logTest('HTTPS connection', false, error.message);
                resolve(false);
            }
        });
    }

    async test_tls_1_2() {
        return new Promise((resolve) => {
            try {
                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com',
                    minVersion: 'TLSv1.2',
                    maxVersion: 'TLSv1.2',
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const protocol = socket.getProtocol();
                    const cipher = socket.getCipher();

                    const success = protocol === 'TLSv1.2';
                    this.logTest('TLS 1.2 connection', success,
                        `Protocol: ${protocol}, Cipher: ${cipher.name}`);

                    socket.end();
                    resolve(success);
                });

                socket.on('error', (error) => {
                    this.logTest('TLS 1.2 connection', false, error.message);
                    resolve(false);
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('TLS 1.2 connection', false, 'Connection timeout');
                    resolve(false);
                });
            } catch (error) {
                this.logTest('TLS 1.2 connection', false, error.message);
                resolve(false);
            }
        });
    }

    async test_tls_1_3() {
        return new Promise((resolve) => {
            try {
                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com',
                    minVersion: 'TLSv1.3',
                    maxVersion: 'TLSv1.3',
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const protocol = socket.getProtocol();
                    const cipher = socket.getCipher();

                    const success = protocol === 'TLSv1.3';
                    this.logTest('TLS 1.3 connection', success,
                        `Protocol: ${protocol}, Cipher: ${cipher.name}`);

                    socket.end();
                    resolve(success);
                });

                socket.on('error', (error) => {
                    // TLS 1.3 may not be supported, treat as acceptable
                    this.logTest('TLS 1.3 connection', true,
                        'Not supported by server (acceptable)');
                    resolve(true);
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('TLS 1.3 connection', false, 'Connection timeout');
                    resolve(false);
                });
            } catch (error) {
                this.logTest('TLS 1.3 connection', false, error.message);
                resolve(false);
            }
        });
    }

    async test_certificate_validation() {
        return new Promise((resolve) => {
            try {
                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com',
                    rejectUnauthorized: true,
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const cert = socket.getPeerCertificate();
                    const authorized = socket.authorized;

                    this.logTest('Certificate validation', authorized,
                        `Subject: ${cert.subject?.CN}, Authorized: ${authorized}`);

                    socket.end();
                    resolve(authorized);
                });

                socket.on('error', (error) => {
                    this.logTest('Certificate validation', false, error.message);
                    resolve(false);
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('Certificate validation', false, 'Connection timeout');
                    resolve(false);
                });
            } catch (error) {
                this.logTest('Certificate validation', false, error.message);
                resolve(false);
            }
        });
    }

    async test_fips_cipher() {
        return new Promise((resolve) => {
            try {
                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com',
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const cipher = socket.getCipher();
                    const cipherName = cipher.name.toLowerCase();

                    const isFips = cipherName.includes('aes') &&
                                  (cipherName.includes('gcm') || cipherName.includes('ccm'));

                    this.logTest('FIPS-approved cipher negotiation', isFips,
                        `Cipher: ${cipher.name}`);

                    socket.end();
                    resolve(isFips);
                });

                socket.on('error', (error) => {
                    this.logTest('FIPS-approved cipher negotiation', false, error.message);
                    resolve(false);
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('FIPS-approved cipher negotiation', false, 'Connection timeout');
                    resolve(false);
                });
            } catch (error) {
                this.logTest('FIPS-approved cipher negotiation', false, error.message);
                resolve(false);
            }
        });
    }

    async test_https_post() {
        return new Promise((resolve) => {
            try {
                const postData = JSON.stringify({ test: 'fips-tls' });

                const options = {
                    hostname: 'httpbin.org',
                    port: 443,
                    path: '/post',
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'Content-Length': Buffer.byteLength(postData)
                    },
                    timeout: 10000
                };

                const req = https.request(options, (res) => {
                    const cipher = res.socket.getCipher();

                    const success = res.statusCode === 200;
                    this.logTest('HTTPS POST request', success,
                        `Status: ${res.statusCode}, Cipher: ${cipher.name}`);

                    let data = '';
                    res.on('data', (chunk) => { data += chunk; });
                    res.on('end', () => resolve(success));
                });

                req.on('error', (error) => {
                    this.logTest('HTTPS POST request', false, error.message);
                    resolve(false);
                });

                req.on('timeout', () => {
                    req.destroy();
                    this.logTest('HTTPS POST request', false, 'Request timeout');
                    resolve(false);
                });

                req.write(postData);
                req.end();
            } catch (error) {
                this.logTest('HTTPS POST request', false, error.message);
                resolve(false);
            }
        });
    }

    async runAllTests() {
        console.log('');
        console.log('Running TLS/SSL Tests...');
        console.log('');

        await this.test_https_connection();
        await this.test_tls_1_2();
        await this.test_tls_1_3();
        await this.test_certificate_validation();
        await this.test_fips_cipher();
        await this.test_https_post();

        console.log('');
        console.log(`TLS Tests: ${this.passedTests}/${this.totalTests} passed`);
        console.log('');

        return this.passedTests >= 5 ? 0 : 1;
    }
}

module.exports = TlsTestSuite;

// Allow running standalone
if (require.main === module) {
    (async () => {
        const suite = new TlsTestSuite();
        const exitCode = await suite.runAllTests();
        process.exit(exitCode);
    })();
}
